### PR TITLE
[no squash] Windows stuff

### PR DESCRIPTION
**Pull request by sfan5**
there isn't much to write here because I tested all of it

perhaps about the second commit: if you run this test with a MinGW build it should fail.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -813,7 +815,7 @@ else()
 	endif()
 
 	if(MINGW)
-		set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -mwindows")
+		set(CMAKE_EXE_LINKER_FLAGS_RELEASE "-mwindows")
```
Comment (lhofhansl) [inline src/CMakeLists.txt:818]: I do not understand this one. What does this change? And why is this only needed for MINGW?

**Comment by lhofhansl** [comment]
Does this change the MINGW crash upon exit?

File: src/CMakeLists.txt (hunk 1 above)
Comment (sfan5) [reply src/CMakeLists.txt:818]: lazily copied from docs:

>  This option is available for Cygwin and MinGW targets. It specifies that a GUI application is to be generated by instructing the linker to set the PE header subsystem type appropriately.

the `if(MSVC)` has it too, but with a different syntax.
it's also only needed at link time, if you use it for compiling files clang will print a warning.

**Comment by sfan5** [comment]
> Does this change the MINGW crash upon exit?

no
if you want to test a build that doesn't crash try this: 
[minetest-llvm-mingw.zip](https://github.com/minetest/minetest/files/13813038/minetest-llvm-mingw.zip)

File: src/porting.cpp [hunk 2]
```diff
@@ -738,10 +738,12 @@ void attachOrCreateConsole()
...
-	if (!consoleAllocated && redirected && (AttachConsole(ATTACH_PARENT_PROCESS) || AllocConsole())) {
-		freopen("CONOUT$", "w", stdout);
-		freopen("CONOUT$", "w", stderr);
+	const bool redirected = _fileno(stdout) >= 0; // If output is redirected to e.g a file
+	if (!consoleAllocated && !redirected) {
+		if (AttachConsole(ATTACH_PARENT_PROCESS) || AllocConsole()) {
+			freopen("CONOUT$", "w", stdout);
+			freopen("CONOUT$", "w", stderr);
+		}
 		consoleAllocated = true;
```
Comment (SmallJoker) [inline src/porting.cpp:747]: At first it seemed like the variable should be in the code block above. Turns out it's a slightly misleading name.
Might you please be so nice to change it to `redirect_attempted`, `run_once`, or similar?

File: src/main.cpp [hunk 3]
```diff
@@ -213,6 +213,7 @@ int main(int argc, char *argv[])
 
 	// Run unit tests
 	if (cmd_args.getFlag("run-unittests")) {
+		porting::attachOrCreateConsole();
```
Comment (SmallJoker) [inline src/main.cpp:216]: Creating the console here is rather useless because the process exits immediately.

**Comment by SmallJoker** [comment]
Did a few quick tests on Windows (Wine is too broken) and the console does show up as expected. `--run-unittests` no longer needs the `--console` argument, which is good to see. Also a minidump file is generated upon crash, which is at least something.

There seems to be an issue after the unittest "testGetModMediaPathsWrongDir". The [most recent](https://github.com/minetest/minetest/actions/runs/7389941735) MSVC x64 build produces the following dump:
```
FAULTING_IP: 
minetest+3edd88
00007ff7`69ebdd88 4c8b7f10        mov     r15,qword ptr [rdi+10h]

EXCEPTION_RECORD:  (.exr -1)
ExceptionAddress: 00007ff769ebdd88 (minetest+0x00000000003edd88)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 0000000000000000
   Parameter[1]: 0000000000000050
Attempt to read from address 0000000000000050


STACK_TEXT:  
00000059`51fef320 00007ff7`69e72066 : 00000059`51fef690 00000167`26006900 00000059`51fef690 00000000`00000300 : minetest+0x3edd88
00000059`51fef5c0 00007ff7`69ec0feb : 00007ff7`6a27ec30 00007ff7`6a15bc78 00007ff7`6a27ec30 00000000`0009611c : minetest+0x3a2066
00000059`51fef630 00007ff7`69e727d4 : 00000000`000960ec 00000000`00000308 00000167`26006900 00007ff7`6a27ec30 : minetest+0x3f0feb
00000059`51fef6e0 00007ff7`69e72ec0 : 00000167`2601c3d0 00000000`00000003 00000000`00000003 00000000`00000000 : minetest+0x3a27d4
00000059`51fef730 00007ff7`69f355c1 : 00000000`000951ce 00000000`00000010 00000000`00000000 00000000`00000000 : minetest+0x3a2ec0
00000059`51fef7c0 00007ff7`69c52edc : 00007ff7`00000002 00000167`2603cf90 00000059`51fefa01 00000167`2605edd0 : minetest+0x4655c1
00000059`51fefde0 00007ffb`b6f774b4 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : minetest+0x182edc
00000059`51fefe20 00007ffb`b83226a1 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : kernel32!BaseThreadInitThunk+0x14
00000059`51fefe50 00000000`00000000 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : ntdll!RtlUserThreadStart+0x21
```

But that can be fixed in a later PR. LGTM.

**Comment by sfan5** [comment]
> this rather prevents me from checking whether `testTLS()` works

You can use `--test-module TestThreading`.

File: src/main.cpp (hunk 3 above)
Comment (sfan5) [reply src/main.cpp:216]: indeed. but if you're running MT from an existing console this is also needed for output to appear there at all

**Comment by SmallJoker** [comment]
> You can use `--test-module TestThreading`.

I figured this requires `--run-unittests` to be specified as well. This is not visible from the `--help` output but fine by me. The tests passed.

**Review by SmallJoker — APPROVED** [review]
Thank you!

**Pull request merged**